crates/base_db/src/input.rs

```diff
@@ -102,11 +102,46 @@ impl fmt::Display for CrateName {
 
 impl ops::Deref for CrateName {
     type Target = str;
-    fn deref(&self) -> &Self::Target {
+    fn deref(&self) -> &str {
         &*self.0
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CrateDisplayName {
+    // The name we use to display various paths (with `_`).
+    crate_name: CrateName,
+    // The name as specified in Cargo.toml (with `-`).
+    canonical_name: String,
```

Comment:
This field doesn't seem to be used.

Reply:
Although I expect we *should* use in some cases. For exaple, link to docs.rs ideally should be `https://docs.rs/expect-test` rather than `https://docs.rs/expect_test`.

Reply:
Hmm. Isn't that `expect-test` the _package_ name? I can never get the terminology straight there...

Reply:
Yes, it is the package name. A package may contain at most one library crate and any amount of binary crates.

Reply:
Yeah, good question, haven't thought about that.

For smth like

```toml
[package]
name = "foo"

[lib]
name = "bar"
```

the crate_name will be foo, and canonical_name will be bar. We'd use `bar` when we need to display a path, and `foo` for doc.rs links. canonical is probably not the best term, I've stolen that from -/_ pre RFC